src/Numerics/DGMethods/FVModel.jl

```diff
@@ -0,0 +1,483 @@
+"Observations on the fifth-order WENO method with non-uniform meshes." 
+Applied Mathematics and Computation 196.1 (2008): 433-447.
+
+size(Δh) = 5
+size(u) = (num_state_primitive, 5)
+construct left/right face states of cell[3]
+h1     h2     h3     h4      h5
+|--i-2--|--i-1--|--i--|--i+1--|--i+2--|
+without hat : i - 1/2
+with hat    : i + 1/2  
+r = 0, 1, 2 cells to the left, => P_r^{i}
+I_{i-r}, I_{i-r+1}, I_{i-r+2}
+P_r(x) =  ∑_{j=0}^{2} C_{rj}(x) u_{i - r + j}  (use these 3 cell averaged values)
+C_{rj}(x) = B_{rj}(x) h_{3-r+j}                (i = 3)
+C''_{rj}(x) = B''_{rj}(x) h_{3-r+j}                (i = 3)
+P''_r(x)    =  ∑_{j=0}^{2} C''_{rj}(x) u_{i - r + j}  (use these 3 cell averaged values)
+=  ∑_{j=0}^{2} B''_{rj}(x) h_{3-r+j}  u_{i - r + j}  (use these 3 cell averaged values)
+
+"""
+
+function weno_reconstruction!(
+    state_primitive_top::Vars,
+    state_primitive_bottom::Vars,
+    cell_states_primitive::NTuple{5, Vars},
+    cell_weights::SVector{5, FT},
+) where {FT}
+
+
+    num_state_primitive = length(parent(state_primitive_top))
+    h1, h2, h3, h4, h5 = cell_weights
+
+    b̂ = zeros(FT, 3, 3)
+    b̂[3, 3] = 1 / (h1 + h2 + h3) + 1 / (h2 + h3) + 1 / h3
+    b̂[3, 2] = b̂[3, 3] - (h1 + h2 + h3) * (h2 + h3) / ((h1 + h2) * h2 * h3)
+    b̂[3, 1] = b̂[3, 2] + (h1 + h2 + h3) * h3 / (h1 * h2 * (h2 + h3))
+    b̂[2, 3] = (h2 + h3) * h3 / ((h2 + h3 + h4) * (h3 + h4) * h4)
+    b̂[2, 2] = b̂[2, 3] + 1 / (h2 + h3) + 1 / h3 - 1 / h4
+    b̂[2, 1] = b̂[2, 2] - ((h2 + h3) * h4) / (h2 * h3 * (h3 + h4))
+    b̂[1, 3] = -(h3 * h4) / ((h3 + h4 + h5) * (h4 + h5) * h5)
+    b̂[1, 2] = b̂[1, 3] + h3 * (h4 + h5) / ((h3 + h4) * h4 * h5)
+    b̂[1, 1] = b̂[1, 2] + 1 / h3 - 1 / h4 - 1 / (h4 + h5)
+
+    b = zeros(FT, 3, 3)
+    b[3, 3] = 1 / (h5 + h4 + h3) + 1 / (h4 + h3) + 1 / h3
+    b[3, 2] = b[3, 3] - (h5 + h4 + h3) * (h4 + h3) / ((h5 + h4) * h4 * h3)
+    b[3, 1] = b[3, 2] + (h5 + h4 + h3) * h3 / (h5 * h4 * (h4 + h3))
+    b[2, 3] = (h4 + h3) * h3 / ((h4 + h3 + h2) * (h3 + h2) * h2)
+    b[2, 2] = b[2, 3] + 1 / (h4 + h3) + 1 / h3 - 1 / h2
+    b[2, 1] = b[2, 2] - ((h4 + h3) * h2) / (h4 * h3 * (h3 + h2))
+    b[1, 3] = -(h3 * h2) / ((h3 + h2 + h1) * (h2 + h1) * h1)
+    b[1, 2] = b[1, 3] + h3 * (h2 + h1) / ((h3 + h2) * h2 * h1)
+    b[1, 1] = b[1, 2] + 1 / h3 - 1 / h2 - 1 / (h2 + h1)
+
+
+    # at i - 1/2, i + 1/2
+    P = zeros(FT, num_state_primitive, 2, 3)
+    for r in 0:2
+        for j in 0:2
+            P[:, 1, 3 - r] +=
+                b[r + 1, j + 1] *
+                cell_weights[3 + r - j] *
+                parent(cell_states_primitive[3 + r - j])
+            P[:, 2, r + 1] +=
+                b̂[r + 1, j + 1] *
+                cell_weights[3 - r + j] *
+                parent(cell_states_primitive[3 - r + j])
+        end
+    end
+
+
+    # build the second derivative part in smoothness measure
+    d2B = zeros(FT, 3, 3)
+    for r in 0:2
+        d2B[r + 1, 3] =
+            6.0 / (
```

Comment:
```suggestion
            6 / (
```
Let's make this type-stable